### PR TITLE
quic: fix build with option `--//bazel:http3=False`

### PR DESCRIPTION
**Pull request by danzh2010**
Commit Message: QUICHE QUIC code and libraries are supposed to be hidden behind macro `ENVOY_ENABLE_QUIC` and `envoy_select_enable_http3()`. Any violation should have resulted in open SSL FIPS build error because of the interface incompatibility. But recent FIPS dependency update https://github.com/envoyproxy/envoy/pull/27087 resolved the incompatibility. So `upstream_request_lib` unhide QUIC code accidentally.  This change fix the dependency leakage and change some critical QUIC libraries to be wrapped in `envoy_select_enable_http3` which skips building the libraries under bazel option `//bazel:http3=False` to avoid future leakage.

Risk Level: low
Testing: existing builds are happy
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Review by adisuissa — DISMISSED** [review]
LGTM, thanks!
Does this PR change the current (user-facing) default support of UDP? If so, it may be good to add a release note.

Assigning Ryan for a senior-maintainer pass.
/assign @RyanTheOptimist

**Comment by adisuissa** [comment]
CI failure seems relevant.
/wait-any

File: source/extensions/upstreams/http/generic/config.cc [hunk 1]
```diff
@@ -28,8 +31,13 @@ Router::GenericConnPoolPtr GenericGenericConnPoolFactory::createGenericConnPool(
         std::make_unique<Upstreams::Http::Tcp::TcpConnPool>(thread_local_cluster, route_entry, ctx);
     return (conn_pool->valid() ? std::move(conn_pool) : nullptr);
   case UpstreamProtocol::UDP:
+#ifdef ENVOY_ENABLE_QUIC
     conn_pool = std::make_unique<Upstreams::Http::Udp::UdpConnPool>(thread_local_cluster, ctx);
```
Comment (RyanTheOptimist) [inline source/extensions/upstreams/http/generic/config.cc:35]: This surprises me slightly. Maybe we need to rethink/rename this class. But "UdpConnPool" doesn't sound like it's QUIC-specific, does it? Should we make this a QuicConnPool?

Actually, looking at source/extensions/upstreams/http/udp/upstream_request.h, it looks like it depends on some QUICHE capsule stuff, but doesn't depend (at least not directly) on QUIC stuff. though for some reason it seems to want to #include quic_spdy_stream.h, but it doesn't seem to use it? I'm confused :)

Comment (danzh2010) [reply source/extensions/upstreams/http/generic/config.cc:35]: @DavidSchinazi, I'm also struggling to fix the leakage from this code. Can you clarify the dependency model here?

Comment (DavidSchinazi) [reply source/extensions/upstreams/http/generic/config.cc:35]: `UdpConnPool` was written by @jeongseokson to add CONNECT-UDP support in #27714. CONNECT-UDP works with HTTP/2 so (in theory) this should work even when HTTP/3 support is compiled out. In practice, that might require a few small refactors to QUICHE because the QUICHE code wasn't designed with the requirement to allow compiling HTTP/3 support out

Comment (danzh2010) [reply source/extensions/upstreams/http/generic/config.cc:35]: QUICHE http2 code doesn't depend on QUICHE quic code. So I *think* we could be able to draw a clear line between http2 code needed by CONNECT-UDP and quic code.

Comment (RyanTheOptimist) [reply source/extensions/upstreams/http/generic/config.cc:35]: Right, that's what I was suggesting. Looking at the *code* it doesn't seem to use QUIC and so we should not need to conditionally compile it. However, for some reason (bug) it does #include quic_spdy_stream.h. I suspect we should just remove that #include?

Comment (jeongseokson) [reply source/extensions/upstreams/http/generic/config.cc:35]: I think @RyanTheOptimist is right, quic_spdy_stream.h. should have not been included in the first place. Sorry for leaving an unncessary include there. Would removing the quic_spdy_stream.h include solve the issue?

File: bazel/external/quiche.BUILD [hunk 2]
```diff
@@ -4247,38 +4249,44 @@ envoy_cc_library(
...
-        "quiche/quic/core/quic_session.h",
-        "quiche/quic/core/quic_stream.h",
-        "quiche/quic/core/quic_stream_id_manager.h",
-        "quiche/quic/core/quic_stream_sequencer.h",
-        "quiche/quic/core/stream_delegate_interface.h",
-        "quiche/quic/core/tls_client_handshaker.h",  # required by tls_handshaker.cc
-        "quiche/quic/core/tls_handshaker.h",
-        "quiche/quic/core/uber_quic_stream_id_manager.h",
-    ],
+    srcs = envoy_select_enable_http3(
```
Comment (RyanTheOptimist) [inline bazel/external/quiche.BUILD:4252]: Is this change needed? It looks like none of the existing QUIC targets in bazel/external/quiche.BUILD are wrapped in envoy_select_enable_http3(). Instead, I think we tend to use envoy_select_enable_http3() up in envoy code.  Should we perhaps we consistent with the other QUIC targets in this file?

**Review by RyanTheOptimist — COMMENTED** [review]
Looking much simpler!

File: bazel/external/quiche.BUILD (hunk 2 above)
Comment (danzh2010) [reply bazel/external/quiche.BUILD:4252]: This will be used to prevent future regression given we can no longer rely on FIPS build to detect unintended leakage. And it seems that only one quic targets using envoy_select_enable_http3() is sufficient instead of all of them.

Comment (RyanTheOptimist) [reply bazel/external/quiche.BUILD:4252]: I like the idea of adding some logic to make sure we don't accidentally leak QUIC code into the no-QUIC build.

If we're going to go this select route, it seems like it would be a good idea to do this for all the various QUIC targets. Wrapping just "quic_core_session_lib" means that it'll break if someone tries to use the code in quic_core_session_lib, or a higher-level target, but it wouldn't break if someone tried to use a *lower* level target (like quic_framer.h, for example). 

What if we did something like:

```
def envoy_quic_cc_library(                                                                                                                                                                          
        name,                                                                                                                                                                                  
        srcs = [],                                                                                                                                                                             
        hdrs = [],                                                                                                                                                                             
        deps = []):                                                                                                                                                                        
    native.cc_library(                                                                                                                                                                         
        name = name,                                                                                                                                                                           
        srcs = envoy_select_enable_http3(srcs),                                                                                                                                                                           
        hdrs = envoy_select_enable_http3(hdrs),                                                                                                                                                                           
        repository = "@envoy",
        copts = quiche_copts,
        tags = ["nofips"],
        visibility = ["//visibility:public"]
        deps = envoy_select_enable_http3(deps),             
    )                                                                                                                                                                                          
```
Then for all the various QUIC targets in this file we could just do:
```
envoy_quic_cc_library(...)
```
And it would automatically do the select for us?

**Review by mattklein123 — APPROVED** [review]

File: bazel/external/quiche.BUILD (hunk 2 above)
Comment (danzh2010) [reply bazel/external/quiche.BUILD:4252]: Adding envoy_quic_cc_library is a good idea. But batch writer also depends on some lower level QUIC libraries (i.e. packet writer interface) and it is supposed to be a non-QUIC extension.  So I guess we have to leave some QUIC libraries unchanged.

Comment (RyanTheOptimist) [reply bazel/external/quiche.BUILD:4252]: Curious. We have some QUIC code that's present in Envoy even when we're not speaking QUIC? Sounds like we should consider moving that code out of QUIC and into QUICHE proper. That being said, in the meantime leaving those libraries unchanged SGTM

Comment (danzh2010) [reply bazel/external/quiche.BUILD]: QuicBatchWriter was implemented with QUIC interfaces, but it can be used for any UDP listeners.

**Review by alyssawilk — APPROVED** [review]
